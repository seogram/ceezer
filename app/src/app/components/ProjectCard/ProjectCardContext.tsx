import { createContext, useContext } from 'react';
import { Project } from '../../type';

const ProjectCardContext = createContext<{ project: Project } | null>(null);

export function useProjectCardContext() {
  const context = useContext(ProjectCardContext);
  if (!context) {
    throw new Error(
      'ProjectCard.* component must be rendered as child of ProjectCard component'
    );
  }
  return context;
}

export default ProjectCardContext;
