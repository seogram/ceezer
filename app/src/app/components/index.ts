export { default as MarketPlace } from "./MarketPlace";
export { default as Loading } from "./Loading/Loading";
export { default as CustomError } from "./Error/Error";
export { default as Header } from "./Header/Header";
export { default as Cart } from "./Cart/Cart";
export { default as CustomeLink } from "./Link/CustomeLink";
export { default as SearchBox } from "./SearchBox/SearchBox";

