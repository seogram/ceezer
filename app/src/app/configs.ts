export const BASE_URL = "https://ceezer-public-assets.s3.eu-central-1.amazonaws.com";
