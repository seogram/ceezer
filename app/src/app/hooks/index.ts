export { useData } from "./useData";
export { useResponsive } from "./useResponsive";
export { default as useCart } from "./useCart";
export { default as useCartActions } from "./useCartActions";

