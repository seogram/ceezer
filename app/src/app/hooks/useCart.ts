import { useContext } from "react";
import { CartItemsContext } from "../context/CartItems";

const useCart = () => {
  const { cartItems, setCartItems } = useContext(CartItemsContext);

  return { cartItems, setCartItems };
};

export default useCart;
