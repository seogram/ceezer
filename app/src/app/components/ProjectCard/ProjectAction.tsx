import * as React from 'react';
import { useProjectCardContext } from './ProjectCardContext';
import { Button, Stack, InputBase } from "@mui/material";
import { Controller, useForm } from "react-hook-form";
import { styled } from "@mui/material/styles";
import { useAddItem } from '../../hooks';

type Volume = {
  volume: number | string;
}

const InputStyle = styled(InputBase)(({ theme }) => ({
  paddingLeft: "10px",
  border: "1px solid",
  width: "10%",
  [theme.breakpoints.down("sm")]: {
    width: "50%",
  },
}));


function ProductAction() {
  const { project } = useProjectCardContext();
  const { addItem } = useAddItem();
  const {
    control,
    setValue,
    handleSubmit,
    watch,
  } = useForm<Volume>({
    mode: "onTouched",
    defaultValues: {
      volume: "",
    },
  });
  
  const volume = Number(watch("volume"));

  const addToCardHandler = () => {
    setValue("volume", "")
    addItem({
      id: project.id,
      name : project.name,
      volume,
      pricePerTon : project.price_per_ton
    })
  }

  return (
    <Stack direction="row" gap={2} sx={{ pt: 2 }} justifyContent={{
      xs: "center",
      sm: "flex-start"
    }}>
      <Controller
        name="volume"
        control={control}
        render={({ field, fieldState: { error } }) => (
          <InputStyle
            {...field}
            autoFocus
            type="number"
            inputProps={{
              min: 0,
              max: project.offered_volume_in_tons
            }}
            placeholder="Volume"
            error={Boolean(error)}
          />
        )}
      />
      <Button
        data-testid="addToCard"
        disabled={!volume || volume === 0}
        variant="contained"
        onClick={
          handleSubmit(addToCardHandler)}
        name="search"
      >
        Add to Card
      </Button>
    </Stack>
  )
}

export default ProductAction;
