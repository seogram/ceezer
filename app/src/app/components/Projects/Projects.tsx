import React from "react";
import { Typography, Button, Box, Stack } from "@mui/material";
import { styled } from "@mui/material/styles";
import { useData } from "../../hooks";
import ProjectDetail from "../ProjectDetail/ProjectDetail";
import Error from "../Error/Error";
import Loading from "../Loading/Loading";
import { useRouter } from "next/navigation";
import { Projects } from "../../type";

const RootStyle = styled(Stack)(() => ({
  gap: "20px",
}));

type Props = {
  searchTerm?: string | null;
};

type useProjectDataReturn = {
  data: Projects;
  isFetching: boolean;
  isError: boolean;
};

const renderDetails = (projects: Projects) =>
  projects.map(project => <ProjectDetail projectData={project} key={project.id} />)


const Projects = ({ searchTerm }: Props) => {
  const router = useRouter();

  const {
    data: projects,
    isFetching: isUserFetching,
    isError,
  }: useProjectDataReturn = useData();

  if (isError) {
    return <Error />;
  }

  if (isUserFetching) {
    return <Loading />;
  }
  const filterByName = () => {
   return projects.filter(project => project.name.search(searchTerm!) > -1);
  }
  
  const filteredProjects = searchTerm ? filterByName() : projects

  return (
    <RootStyle>
      {searchTerm && (
        <Box sx={{ display: "flex", }}>
          <Box>
            <Typography data-test-id="result">
              {`Showing bicycles for  ${searchTerm} :`}
            </Typography>
          </Box>
          <Box sx={{ marginLeft: "auto" }}>
            <Button name="reset" onClick={() => router.replace("/")}>
              Reset Result
            </Button>
          </Box>
        </Box>
      )}
      <Box>
        {renderDetails(filteredProjects)}
      </Box>
    </RootStyle>
  );
};

export default Projects;