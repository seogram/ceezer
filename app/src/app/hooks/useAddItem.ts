import useCart from "./useCart";
// import { IProduct } from "../external/product";
import { setLocalStorage } from "../utils";

const useAddItem = () => {
  const { cartItems, setCartItems } = useCart();

  const addItem = (product: any) => {
    const currentCartItems = [...cartItems];
    const existingCartItem = currentCartItems.find(
      (item) => item.product.id === product.id
    );

    // If the product is already in the cart, update the quantity
    if (existingCartItem) {
      existingCartItem.quantity += 1;
    } else {
      currentCartItems.push({
        product,
        quantity: 1,
      });
    }

    // Update the cart items
    setLocalStorage("cart" , currentCartItems)
    setCartItems(currentCartItems);

  };

  return { addItem };
};

export default useAddItem;