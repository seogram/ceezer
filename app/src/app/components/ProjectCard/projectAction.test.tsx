import { render, fireEvent, waitFor, act, cleanup } from "@testing-library/react";
import ProjectAction from "./ProjectAction";
import { useCart, useCartActions } from "@/app/hooks";
import { MockData, MockCartItemData } from "@/app/components/Projects/data";
import ProjectCardContext from '@/app/context/ProjectCardContext';


jest.mock("../../hooks/useCart");
jest.mock("../../hooks/useCartActions");

const mockSetCartItems = jest.fn();
const mockAddItem = jest.fn();

const mockProject = MockData[0];

const updatedCartAfterIncrease = {
  "id": "1",
  "image": "https://ceezer-public-assets.s3.eu-central-1.amazonaws.com/project_type_sample_images/Fugitives/38bb530f5caf513be9f2a41f2d909f47-min.jpeg",
  "name": "EverGreen CarbonScape",
  "offeredVolume": 15,
  "pricePerTon": 650,
  "volume": 1
}

function setup() {
  (useCart as jest.Mock).mockReturnValue({
    cartItems: MockCartItemData.cartItems,
    setCartItems: mockSetCartItems
  });

  (useCartActions as jest.Mock).mockReturnValue({
    addItem: mockAddItem
  });
  const utils = render(
    <ProjectCardContext.Provider value={{ project: mockProject }}>
      <ProjectAction />
    </ProjectCardContext.Provider>
  );
  return {
    ...utils,
  };
}
describe("ProjectAction Component", () => {

  beforeEach(() => {
    (useCart as jest.Mock).mockReturnValue({
      cartItems: MockCartItemData.cartItems,
      setCartItems: mockSetCartItems
    });

    (useCartActions as jest.Mock).mockReturnValue({
      addItem: mockAddItem
    });
    
  });

  afterEach(cleanup);


  it('renders without crashing', () => {
    const { asFragment } = setup();
    expect(asFragment()).toMatchSnapshot();
  });


  it("User can enter required volume in input box", async () => {
    const { getByPlaceholderText, getByText } = setup();

    const inputElement = getByPlaceholderText("Volume");
    const buttonElement = getByText("Add to Card");

    act(() => {
      fireEvent.change(inputElement, { target: { value: 1 } });
    });

    expect(buttonElement).not.toBeDisabled();

    act(() => {
      fireEvent.click(buttonElement);
    });

    await waitFor(() => {
      expect(mockAddItem).toHaveBeenCalledWith(
        updatedCartAfterIncrease
      );
    });
  });

  it("User can enter required volume in input box", async () => {
    const { getByPlaceholderText, getByText } = setup();

    const inputElement = getByPlaceholderText("Volume");
    const buttonElement = getByText("Add to Card");

    act(() => {
      fireEvent.change(inputElement, { target: { value: 16 } });
    });

    expect(buttonElement).toBeDisabled();
  });
});
