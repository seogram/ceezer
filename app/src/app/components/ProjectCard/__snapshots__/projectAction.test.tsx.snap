// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectAction Component renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="MuiStack-root css-csv01u-MuiStack-root"
  >
    <div
      class="MuiInputBase-root MuiInputBase-colorPrimary css-1t9f0n0-MuiInputBase-root"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input css-yz9k0d-MuiInputBase-input"
        max="15"
        min="0"
        name="volume"
        placeholder="Volume"
        type="number"
        value=""
      />
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-sghohy-MuiButtonBase-root-MuiButton-root"
      disabled=""
      name="search"
      tabindex="-1"
      type="button"
    >
      Add to Card
    </button>
  </div>
  <div
    class="MuiStack-root css-nen11g-MuiStack-root"
  />
</DocumentFragment>
`;
